csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs

```diff
@@ -0,0 +1,179 @@
+            }
+            finally
+            {
+                if (typeAndShape != IntPtr.Zero)
+                {
+                    NativeMethods.OrtReleaseTensorTypeAndShapeInfo(typeAndShape);
+                }
+            }
+
+            switch (elemType)
+            {
+                case TensorElementType.Float:
+                    result = NameOnnxValueFromNativeTensor<float>(name, nativeOnnxValue);
+                    break;
+                case TensorElementType.Double:
+                    result = NameOnnxValueFromNativeTensor<double>(name, nativeOnnxValue);
+                    break;
+                case TensorElementType.Int16:
+                    result = NameOnnxValueFromNativeTensor<short>(name, nativeOnnxValue);
+                    break;
+                case TensorElementType.UInt16:
+                    result = NameOnnxValueFromNativeTensor<ushort>(name, nativeOnnxValue);
+                    break;
+                case TensorElementType.Int32:
+                    result = NameOnnxValueFromNativeTensor<int>(name, nativeOnnxValue);
+                    break;
+                case TensorElementType.UInt32:
+                    result = NameOnnxValueFromNativeTensor<uint>(name, nativeOnnxValue);
+                    break;
+                case TensorElementType.Int64:
+                    result = NameOnnxValueFromNativeTensor<long>(name, nativeOnnxValue);
+                    break;
+                case TensorElementType.UInt64:
+                    result = NameOnnxValueFromNativeTensor<ulong>(name, nativeOnnxValue);
+                    break;
+                case TensorElementType.UInt8:
+                    result = NameOnnxValueFromNativeTensor<byte>(name, nativeOnnxValue);
+                    break;
+                default:
+                    throw new NotSupportedException("Tensor of element type: " + elemType + " is not supported");
+
+            }
+
+            return result;
+        }
+
+
+        private static DisposableNamedOnnxValue NameOnnxValueFromNativeTensor<T>(string name, IntPtr nativeOnnxValue)
+        {
+            NativeOnnxTensorMemory<T> nativeTensorWrapper = new NativeOnnxTensorMemory<T>(nativeOnnxValue);
+            DenseTensor<T> dt = new DenseTensor<T>(nativeTensorWrapper.Memory, nativeTensorWrapper.Dimensions);
+            return new DisposableNamedOnnxValue(name, dt, nativeTensorWrapper);
+        }
+
+
+
+        #region IDisposable Support
+        private bool disposedValue = false; // To detect redundant calls
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposedValue)
+            {
+                if (disposing)
+                {
+                    // TODO: dispose managed state (managed objects).
+                }
+
+                // free unmanaged resources (unmanaged objects) and override a finalizer below.
+                // set large fields to null.
+                if (_nativeMemoryManager != null)
```

Comment:
`_nativeMemoryManager` is a managed object. You can't/shouldn't touch managed objects outside of `if (disposing)`.  This chunk of code should move into the `if (disposing)` block above.